src/Vault.sol

```diff
@@ -0,0 +1,261 @@
+        // Initialize the vault
+        __ERC20_init(name, symbol);
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.paused = true;
+    }
+
+    // QUESTION: Measure the gas difference between IERC20 or address when casting / saving to storage
+    function setRateProvider(address rateProvider_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.rateProvider = rateProvider_;
+        emit SetRateProvider(rateProvider_);
+    }
+
+    function addAsset(address asset_, uint8 decimals_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        if (asset_ == address(0)) revert ZeroAddress();
+        if (decimals_ > 18) revert InvalidDecimals();
+
+        AssetStorage storage assetStorage = Storage.getAssetStorage();
+
+        if (assetStorage.assets[asset_].decimals != 0) revert InvalidAsset();
+
+        uint256 newIndex = assetStorage.list.length;
+
+        assetStorage.assets[asset_] =
+            AssetParams({active: true, index: newIndex, decimals: decimals_, idleAssets: 0, deployedAssets: 0});
+
+        assetStorage.list.push(asset_);
+
+        emit AddAsset(asset_, decimals_, newIndex);
+    }
+
+    function toggleAsset(address asset_, bool active) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        AssetStorage storage assetStorage = Storage.getAssetStorage();
+        if (assetStorage.assets[asset_].decimals == 0) revert AssetNotFound();
+        assetStorage.assets[asset_].active = active;
+        emit ToggleAsset(asset_, active);
+    }
+
+    function pause(bool paused_) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        VaultStorage storage vaultStorage = Storage.getVaultStorage();
+        vaultStorage.paused = paused_;
+        emit Pause(paused_);
+    }
+
+    // QUESTION: What params should be used here? strategy, asset, etc.
+    function processAccounting() public {
+        // get the balances of the assets
+        // AssetStorage storage assetStorage = _getAssetStorage();
+
+        // for (uint256 i = 0; i < assetsStorage.list.length; i++) {
+        //     address asset = assets[i];
+        //     idleBalance = asset.balanceOf(address(vault));
+        // }
+        // get the balances of the strategies
+
+        // call convertToAssets on the strategie?
+
+        // QUESTION: Keep the balances for the assets, or keep that balances in base price
+
+        // NOTE: Get the loops out of the public calls
+
+        //yv3
+        // what if you want to udpate a single strategy, but what if you want to update one strat
+        // the debt value is used to separate the deposited value, you know what the rewards
+        // balance is based on the the debt less the the rewards.
+        // allows you to not have to grind through everything??
```

Comment:
Other things to consider here:

Suppose we have strategies with asynchronous withdrawals:

ynETH
https://etherscan.io/address/0x0BC9BC81aD379810B36AD5cC95387112990AA67b

Example: weETH (hypothetical)
https://etherscan.io/address/0x7d5706f6ef3F89B3951E23e557CDFBC3239D4E2c

What would be an approach to handle totalAssets assuming one of these positions is being unwinded?

Since across withdrawal mechanisms a standard does not exist.

Some may not use NFTs and simply positions so the behaviour is custom.

How. can we code that behaviour to still measure totalAssets?

Reply:
Will pick up this Q in the ynETH scenario test